File: data-structures/Recursion/R_16-MColoringProblem.cpp
#include<bits/stdc++.h>

using namespace std;

class Solution {
  public:
    bool possible(int c, int node, unordered_map<int, vector<int>> adList, vector<int> color){
        for(auto nbr: adList[node]){
            if(color[nbr] == c){
                return false;
            }
        }
        return true;
    }
    bool f(int node, unordered_map<int, vector<int>> &adList, vector<int> &color, 
    int m, int V){
        if(node == V){
            return true;
        }
        for(int c=1; c<=m; c++){
            if(possible(c, node, adList, color)){
                color[node] = c;
                if(f(node+1, adList, color, m, V)){
                    return true;
                }
                color[node] = 0;
            }
        }
        return false;
    }
    bool graphColoring(int v, vector<vector<int>> &edges, int m) {
        // code here
        unordered_map<int, vector<int>> adList;
        for(auto val: edges){
            adList[val[0]].push_back(val[1]);
            adList[val[1]].push_back(val[0]);
        }
        vector<int> color(v, 0);
        int node = 0;
        return f(node, adList, color, m, v);
    }
};

File: data-structures/Recursion/R_16-MColoringProblem.py
# User function Template for python3
from collections import defaultdict
class Solution:
    def graphColoring(self, v, edges, m):
        # code here
        color = [0 for _ in range(v)]
        adList = defaultdict(list)
        for x in edges:
            adList[x[0]].append(x[1])
            adList[x[1]].append(x[0])
        
        def solve(node):
            if(node == v):
                return True
            for col in range(1, m+1):
                if(isSafe(col, node)):
                    color[node] = col
                    if(solve(node+1)):
                        return True
                    color[node] = 0
            return False
        def isSafe(col, node):
            for nbr in adList[node]:
                if color[nbr] == col:
                    return False
            return True
        
        return solve(0)